lib/puppet/provider.rb

```diff
@@ -346,13 +355,17 @@ def self.specificity
     # complexity of a provider).
     match = default_match
     length = match ? match.length : 0
+
+    #TODO: should notdefaultfor affect specificity?
```

Comment:
`notdefaultfor` already does affect specificity, since it affects what comes back from `match = default_match`. It seems fine that it is, since `defaultfor` was already affecting it, and this is effectively refining how `defaultfor` works.

Reply:
Unfortunately, I think the only place that specificity is used is (here)[https://github.com/puppetlabs/puppet/blob/5f4e2804521b39757057520553cf7871f3010191/lib/puppet/type.rb#L1767:L1768], which is after default is already selected, so the match.length can only be for defaultfor, not for notdefaultfor.